### ISSUE TITLE
PyStruct is BSD-licensed but is inoperable without GPL-licensed CVXOPT

### ISSUE THREAD

**Issue by BenKaehler** [pystruct/pystruct#192]
I was hoping to use PyStruct for a BSD-licensed project, but when I try to run

```
from pystruct.learners import SubgradientSSVM
```

which apparently does not rely on `cvxopt`, I get

```
ImportError: No module named 'cvxopt'
```

This is a show stopper for me. Is there any possibility that this dependency might be made optional at some point in the near future?

Thank you!

**Comment by BenKaehler** [comment]
Also, AD3 is a required dependency. Is there any chance of that changing any time soon?

Thanks again.

**Comment by amueller** [comment]
Neigher of them are required dependencies.
You can do
 ``from pystruct.learners.subgradient import SubgradientSSVM``
and you won't wee the cvxopt error.

AD3 should not be required anywhere.

**Comment by BenKaehler** [comment]
Thanks very much @amueller.

When I run 
```
from pystruct.learners.subgradient import SubgradientSSVM
```

I get 
```
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
<ipython-input-2-2805ab9ce42b> in <module>()
----> 1 from pystruct.learners.subgradient import SubgradientSSVM

/Users/ben/miniconda2/envs/qiime2/lib/python3.5/site-packages/pystruct/learners/__init__.py in <module>()
----> 1 from .n_slack_ssvm import NSlackSSVM
      2 from .subgradient_ssvm import SubgradientSSVM
      3 from .downhill_simplex_ssvm import PrimalDSStructuredSVM
      4 from .structured_perceptron import StructuredPerceptron
      5 from .one_slack_ssvm import OneSlackSSVM

/Users/ben/miniconda2/envs/qiime2/lib/python3.5/site-packages/pystruct/learners/n_slack_ssvm.py in <module>()
     10 
     11 import numpy as np
---> 12 import cvxopt
     13 import cvxopt.solvers
     14 

ImportError: No module named 'cvxopt'
```

Also, `ad3` is an `install_requires` in `setup.py`, so whether you need it or not it it gets installed.

**Comment by amueller** [comment]
Yeah I wanted to remove that from the setup.py. You can do that locally or tell pip not to install the dependencies, right?

It's weird that the ``__init__.py`` get's called if you do the direct import. It's pretty easy to make the import lazy in the NSlackSSVM and OneSlackSSVM, though.

**Comment by amueller** [comment]
I don't have a lot of time to work on pystruct right now put PR welcome.

**Comment by BenKaehler** [comment]
Thanks again. If structured learning turns out to give significant improvements over what's available in scikit-learn for my problem you will be hearing from me.

**Comment by amueller** [comment]
cool. Can I ask what you're working on?

**Comment by BenKaehler** [comment]
Sure! It's [here](https://github.com/BenKaehler/q2-feature-classifier). It's a core plugin for [QIIME 2](http://2.qiime.org/docs/index.html). QIIME 2 is the successor to [QIIME](http://www.nature.com/nmeth/journal/v7/n5/full/nmeth.f.303.html), a microbiome analysis framework.
